src/dev.rs

```diff
@@ -689,15 +752,59 @@ impl<F: FieldExt> MockProver<F> {
-                                Value::Real(_) => Some(VerifyFailure::ConstraintNotSatisfied {
-                                    constraint: (
-                                        (gate_index, gate.name()).into(),
-                                        poly_index,
-                                        gate.constraint_name(poly_index),
-                                    )
-                                        .into(),
-                                    row: (row - n) as usize,
-                                }),
+                                Value::Real(_) => {
+                                    // Collect cell values.
+                                    let virtual_cells = gate.queried_cells();
+                                    let cell_values = poly.evaluate(
+                                        &|_| BTreeMap::default(),
+                                        &|_| {
+                                            panic!(
+                                                "virtual selectors are removed during optimization"
+                                            )
+                                        },
+                                        &cell_value(
+                                            virtual_cells,
+                                            Any::Fixed,
+                                            &load(n, row, &self.cs.fixed_queries, &self.fixed),
+                                        ),
+                                        &cell_value(
+                                            virtual_cells,
+                                            Any::Advice,
+                                            &load(n, row, &self.cs.advice_queries, &self.advice),
+                                        ),
+                                        &cell_value(
+                                            virtual_cells,
+                                            Any::Instance,
+                                            &load_instance(
+                                                n,
+                                                row,
+                                                &self.cs.instance_queries,
+                                                &self.instance,
+                                            ),
+                                        ),
+                                        &|a| a,
+                                        &|mut a, mut b| {
+                                            a.append(&mut b);
+                                            a
+                                        },
+                                        &|mut a, mut b| {
+                                            a.append(&mut b);
+                                            a
+                                        },
+                                        &|a, _| a,
+                                    );
```

Comment:
Again this is a lot of code to be adding here.

Reply:
Extracted this as a utility function.